presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java

```diff
@@ -23,79 +24,190 @@
 import com.facebook.presto.sql.tree.LongLiteral;
 import com.facebook.presto.sql.tree.SingleColumn;
 import com.facebook.presto.verifier.framework.Column;
+import com.facebook.presto.verifier.framework.VerifierConfig;
 import com.google.common.collect.ImmutableList;
 
+import javax.inject.Inject;
+
 import java.util.List;
 import java.util.Objects;
 import java.util.Optional;
 
 import static com.facebook.presto.sql.QueryUtil.functionCall;
 import static com.facebook.presto.verifier.checksum.ArrayColumnValidator.generateArrayChecksum;
+import static com.facebook.presto.verifier.checksum.ArrayColumnValidator.getAsDoubleArrayColumn;
 import static com.facebook.presto.verifier.framework.VerifierUtil.delimitedIdentifier;
 import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
 
 public class MapColumnValidator
         implements ColumnValidator
 {
+    private final FloatingPointColumnValidator floatingPointValidator;
+    private final boolean validateStringAsDouble;
+
+    @Inject
+    public MapColumnValidator(VerifierConfig config, FloatingPointColumnValidator floatingPointValidator)
+    {
+        this.floatingPointValidator = requireNonNull(floatingPointValidator, "floatingPointValidator is null");
+        this.validateStringAsDouble = config.isValidateStringAsDouble();
+    }
+
     @Override
     public List<SingleColumn> generateChecksumColumns(Column column)
     {
         checkArgument(column.getType() instanceof MapType, "Expect MapType, found %s", column.getType().getDisplayName());
         Type keyType = ((MapType) column.getType()).getKeyType();
         Type valueType = ((MapType) column.getType()).getValueType();
+        ImmutableList.Builder<SingleColumn> builder = ImmutableList.builder();
 
         Expression checksum = functionCall("checksum", column.getExpression());
         Expression keysChecksum = generateArrayChecksum(functionCall("map_keys", column.getExpression()), new ArrayType(keyType));
+        // checksum(cardinality(map_column))
         Expression mapCardinalityChecksum = functionCall("checksum", functionCall("cardinality", column.getExpression()));
+        // coalesce(sum(cardinality(map_column)), 0)
         Expression mapCardinalitySum = new CoalesceExpression(
                 functionCall("sum", functionCall("cardinality", column.getExpression())),
                 new LongLiteral("0"));
 
+        builder.add(new SingleColumn(checksum, Optional.of(delimitedIdentifier(getChecksumColumnAlias(column)))));
+        builder.add(new SingleColumn(keysChecksum, Optional.of(delimitedIdentifier(getKeysChecksumColumnAlias(column)))));
+
         // We need values checksum in one case only: when key is a floating point type and value is not.
```

Comment:
If we're doing floating point validation on the map now does this still apply? wouldn't we always want to do floating point validation on the values if they are floating point types?

Reply:
Nope, for map of floating points, we're not doing floating points validation today. And I think we need to add that. I see you discussed about this in https://github.com/prestodb/presto/pull/22793#issuecomment-2127656021. And I agree. Either me or Sergey could add that later. But it needs a separate PR, given this PR is already very large.

Reply:
I think I'm a bit confused because we are generating all the columns for floating point values in maps as we would for floating point validation.  Is the difference that we are then comparing that they are exactly equal rather than using an error margin?

Reply:
Hi Rebecca, for the branch of map of floating point, we're not generating columns for floating point validation today.   And I keep it as it is with this PR. We do the checksum exact match. So this MapColumnValidator component probably often fails I guess. (Attaching the logic block below) There's another component following this ColumnValidator component in the verifier workflow, called FailureResolver. For map mismatch returned by ColumnValidator, if it sees the map has a floating point key, the FailureResolver auto resolves the mismatches and considers it correct.

I think this MapColumnValidator + FailureResolver for map is too inaccurate. And we should do floating point validation in MapColumnValidator, and get rid of FailureResolve. The recent https://github.com/prestodb/presto/pull/22143 improved ArrayColumnValidator to do floating point validation for array. And we should make the same change for map. That needs a separate PR.

Then in this PR which is for map of varchar, I make the path for map of varchar directly do floating point validation. We're generating floating point columns only for this path.

https://github.com/prestodb/presto/blob/7dae0d6f98ecba6a0b6edd1c073f0b4cbdba460d/presto-verifier/src/main/java/com/facebook/presto/verifier/checksum/MapColumnValidator.java#L73-L82

https://github.com/prestodb/presto/blob/f16032e9e395afc4576eea0337c74cbee3920569/presto-verifier/src/main/java/com/facebook/presto/verifier/resolver/StructuredColumnMismatchResolver.java#L76-L77

Reply:
I see so am I correct that with your change if it's a double string we'll use floating point validation if the config property is set, but if it's just a double we don't do floating point validation?  Can we have floating point types go through the same path?

Reply:
> I see so am I correct that with your change if it's a double string we'll use floating point validation if the config property is set, but if it's just a double we don't do floating point validation?

@rschlussel Rebecca, yes, exactly, that is what this PR has.

Regarding having the true floating point type go through the same path, yes, we could and should do that next. The floating point validation is abstracted into a method in FloatingPointColumnValidator. And we could call that, after refactoring a bit of this MapColumnValidator.validate(). We prefer to make it a separate PR. These changes need quite a bit testing to make sure free of false positive and negative. I would like to separate the change.